### PR TITLE
Cassandra 16446

### PR DESCRIPTION
**Pull request by bereng**

### CODE REVIEW

**Comment by bereng** [comment]
- CI [j11](https://app.circleci.com/pipelines/github/bereng/cassandra/217/workflows/5dda8acf-5214-4444-bdde-7e780527d703)
- Ci [j8](https://app.circleci.com/pipelines/github/bereng/cassandra/217/workflows/699173b0-a40c-4926-b99e-d44160d3127d)

Known offenders or pass locally. Lgtm.

File: src/java/org/apache/cassandra/service/ActiveRepairService.java [hunk 1]
```diff
@@ -304,6 +306,12 @@ public int getRepairSessionSpaceInMegabytes()
         return stats;
     }
 
+    @Override
+    public int parentRepairSessionCount()
```
Comment (ekaterinadimitrova2) [inline src/java/org/apache/cassandra/service/ActiveRepairService.java]: nit which I understand is not part of this patch and you override the method but I would personally go ahead and change it to `parentRepairSessionsCount()`

File: src/java/org/apache/cassandra/service/ActiveRepairServiceMBean.java [hunk 2]
```diff
@@ -41,4 +41,13 @@
...
     public List<CompositeData> cleanupPending(List<String> schemaArgs, String rangeString, boolean force);
+
+    /**
+     * Each ongoing repair (incremental and non-incremental) is represented by a
+     * {@link ActiveRepairService.ParentRepairSession} entry in the {@link ActiveRepairService} cache.
+     * Returns the current number of ongoing repairs (the current number of cached entries).
+     *
+     * @return current size of the internal cache holding {@link ActiveRepairService.ParentRepairSession} instances
+     */
+    int parentRepairSessionCount();
```
Comment (ekaterinadimitrova2) [inline src/java/org/apache/cassandra/service/ActiveRepairServiceMBean.java]: I think we should mention this exposure in the ticket

File: src/java/org/apache/cassandra/repair/RepairRunnable.java [hunk 3]
```diff
@@ -393,22 +393,23 @@ private void repair(String[] cfnames, NeighborsAndRanges neighborsAndRanges)
     {
         if (options.isPreview())
         {
-            previewRepair(parentSession, creationTimeMillis, neighborsAndRanges.filterCommonRanges(keyspace, cfnames), cfnames);
+            previewRepair(parentSession, creationTimeMillis, neighborsAndRanges.filterCommonRanges(keyspace, cfnames), neighborsAndRanges.participants ,cfnames);
```
Comment (ekaterinadimitrova2) [inline src/java/org/apache/cassandra/repair/RepairRunnable.java]: Correct space: `previewRepair(parentSession, creationTimeMillis, neighborsAndRanges.filterCommonRanges(keyspace, cfnames), neighborsAndRanges.participants, cfnames);`

File: src/java/org/apache/cassandra/repair/RepairRunnable.java [hunk 4]
```diff
@@ -393,22 +393,23 @@ private void repair(String[] cfnames, NeighborsAndRanges neighborsAndRanges)
...
-            normalRepair(parentSession, creationTimeMillis, traceState, neighborsAndRanges.filterCommonRanges(keyspace, cfnames), cfnames);
+            normalRepair(parentSession, creationTimeMillis, traceState, neighborsAndRanges.filterCommonRanges(keyspace, cfnames), neighborsAndRanges.participants, cfnames);
         }
     }
 
     private void normalRepair(UUID parentSession,
                               long startTime,
                               TraceState traceState,
                               List<CommonRange> commonRanges,
+                              Set<InetAddressAndPort> cleanupEndpoints,
```
Comment (ekaterinadimitrova2) [inline src/java/org/apache/cassandra/repair/RepairRunnable.java]: I would keep it `preparedEndpoints`

File: src/java/org/apache/cassandra/repair/RepairRunnable.java [hunk 5]
```diff
@@ -699,6 +706,7 @@ public void onSuccess(Object result)
             else
             {
                 success("Repair completed successfully");
+                ActiveRepairService.instance.cleanUp(parentSession, cleanupEndpoints);
```
Comment (ekaterinadimitrova2) [inline src/java/org/apache/cassandra/repair/RepairRunnable.java]: Why not remove the parentSession too?
`ActiveRepairService.instance.removeParentRepairSession(parentSession);`
Is it considered change of behavior? 
/CC @adelapena

Comment (bereng) [reply src/java/org/apache/cassandra/repair/RepairRunnable.java]: Parent session is removed as part of the `success()` call path.

Comment (ekaterinadimitrova2) [reply src/java/org/apache/cassandra/repair/RepairRunnable.java]: My bad, found it in the complete method, looks like I missed it last night. Thanks

File: src/java/org/apache/cassandra/repair/RepairRunnable.java [hunk 6]
```diff
@@ -393,22 +393,23 @@ private void repair(String[] cfnames, NeighborsAndRanges neighborsAndRanges)
     {
         if (options.isPreview())
         {
-            previewRepair(parentSession, creationTimeMillis, neighborsAndRanges.filterCommonRanges(keyspace, cfnames), cfnames);
+            previewRepair(parentSession, creationTimeMillis, neighborsAndRanges.filterCommonRanges(keyspace, cfnames), neighborsAndRanges.participants, cfnames);
```
Comment (adelapena) [inline src/java/org/apache/cassandra/repair/RepairRunnable.java]: Nit: maybe we could break this line, and also the two analogous lines after:
```java
previewRepair(parentSession,
              creationTimeMillis,
              neighborsAndRanges.filterCommonRanges(keyspace, cfnames),
              neighborsAndRanges.participants,
              cfnames);
```

File: src/java/org/apache/cassandra/repair/RepairRunnable.java [hunk 7]
```diff
@@ -447,13 +448,14 @@ public ListenableFuture apply(List<RepairSessionResult> results)
                 return Futures.immediateFuture(null);
             }
         }, MoreExecutors.directExecutor());
-        Futures.addCallback(repairResult, new RepairCompleteCallback(parentSession, successfulRanges, startTime, traceState, hasFailure, executor), MoreExecutors.directExecutor());
+        Futures.addCallback(repairResult, new RepairCompleteCallback(parentSession, successfulRanges, preparedEndpoints, startTime, traceState, hasFailure, executor), MoreExecutors.directExecutor());
```
Comment (adelapena) [inline src/java/org/apache/cassandra/repair/RepairRunnable.java]: These are also getting quite long, perhaps we could break them:
```java
Futures.addCallback(repairResult,
                    new RepairCompleteCallback(parentSession, successfulRanges, preparedEndpoints, startTime, traceState, hasFailure, executor),
                    MoreExecutors.directExecutor());
```

**Review by adelapena — APPROVED** [review]
Looks good to me, +1 assuming CI looks good.

**Review by ekaterinadimitrova2 — APPROVED** [review]
Rebase and green CI and I think this is good to go, thanks!

**Comment by bereng** [comment]
Given @ekaterinadimitrova2 has already rebased and has a green CI I would decline the 2 nits. Otherwise I would have done them.

File: src/java/org/apache/cassandra/repair/RepairRunnable.java (hunk 6 above)
Comment (bereng) [reply src/java/org/apache/cassandra/repair/RepairRunnable.java]: Given @ekaterinadimitrova2 has already rebased and has a green CI I would decline the 2 nits. Otherwise I would have done them.

File: src/java/org/apache/cassandra/repair/RepairRunnable.java (hunk 7 above)
Comment (bereng) [reply src/java/org/apache/cassandra/repair/RepairRunnable.java]: Given @ekaterinadimitrova2 has already rebased and has a green CI I would decline the 2 nits. Otherwise I would have done them.

Comment (ekaterinadimitrova2) [reply src/java/org/apache/cassandra/repair/RepairRunnable.java]: I think there is some confusion here, I just used your branch to start Jenkins CI run. No rebase or squash was done on my end :-) That is the difference with CircleCI, in Jenkins you can initiate run with anyone's fork branches. No need to be your own one

File: src/java/org/apache/cassandra/repair/RepairRunnable.java (hunk 6 above)
Comment (bereng) [reply src/java/org/apache/cassandra/repair/RepairRunnable.java]: Done

File: src/java/org/apache/cassandra/repair/RepairRunnable.java (hunk 7 above)
Comment (bereng) [reply src/java/org/apache/cassandra/repair/RepairRunnable.java]: Done

**Comment by bereng** [comment]
Formatting of those few long lines pushed. Ready to commit now imo?
